accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,341 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+  }
+
+  { // late and/or final the same
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    } else {
+      x = null;
+    } // x has type int?
+  }
+
+  {
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // x has type int?
+  }
+
+  {
+    final x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // Error: not assigned on all paths
+  }
+
+  { // late final the same
+    late x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // x has type int
+  }
+
+  var z; // z is an untyped variable
+  if (b) {
+    z = 3; // z is promoted to int
+   } else {
+    z = 3.0; // z is promoted to double
+    return;
+  } // No error.  z is inferred as int on the only reachable path
+  z.isEven; // No error, only one reachable path
+
+  var u; // u is an untyped variable
+  if (b) {
+    u = 0 as num; // u is promoted to num
+    if (u is! int) u = 1;
+    // u has inferred type num, but is promoted to int
+  } else {
+    u = 3;
+  } // Error: u is inferred to different types on different paths
+}
+```
+
+It is an error to read an **untyped variable** when it has no **initialization
+inferred type**.
+
+
+```dart
+void test(bool b) {
+  {
+    var x; // x is an untyped variable
+    print(x); // Error, no initialization inferred type
+  }
+
+  {
+    late var x; // x is an untyped variable
+    var f = () => print(x); // Error, x has no initialization inferred type
+  }
+
+  {
+    var x; // x is an untyped variable
+    if (b) {
+      print(x); // Error: x has no initialization inferred type.
+      x = 3;
+    } else {
+      x = 4;
+    } // x is inferred as type int
+    print(x); // No error here, since x has been inferred at this point
+  }
+
+  {
+    var z;
+    do {
+      // z is potentially assigned.
+      if (b) {
+        z = 42;
+      } else {
+        z = 37;
+      }
+    } while (something);
+    print(z); // Error, z has no initialization inferred type
+  }
+```
+
+As a consequence of the above it is impossible, in an error-free program, to
+observe the static type or contents of an untyped variable before it has been
+given an inferred type.  From this standpoint it may be thought of as having no
+type.  However, IDEs and tools may wish to report a static type for such a
+variable for the purposes of documentation and error reporting.  For such
+purposes, tools shall treat an untyped variable as if it were declared with a
+type equal to the upper bound of all of the **initialization inferred types**
+that it is ascribed in the method, plus `Never`.  The latter implies that a
+variable which is ascribed no **initialization inferred types** on any path
+shall be treated as having declared type `Never`.  Note that despite the
+requirement that the **initialization inferred types** be equal at join points,
+it is possible for a variable to take on different **initialization inferred
+types** on paths that never join before exiting.
```

Comment:
So then a stranger example is:

```dart
test(c) {
  var x;
  if (c) {
    x = "a string";
    return;
  } else {
    x = 123;
    return;
  }
}
```

So you can create a variable and use it as totally arbitrary unrelated types as long as you escape before those paths join. It's a little strange to me that the above example is allowed but this is not:

```dart
test(c) {
  var x;
  if (c) {
    x = "a string";
  } else {
    x = 123;
  }
}
```

With `final`, we don't make it an error to join two paths where one is definitely assigned and the other is not. We just leave the variable in a half-assigned unusable state and then make it an error to touch it. But here we say the join itself is the point of error.

Should we pick one or the other strategy and have a more consistent model?

I lean slightly towards making the use of the variable the error and not the join itself. Pragmatically, that allows more programs that don't do anything harmful. And I think it draws the user's attention to a concrete entity — a variable they declared that they are using — instead of to a somewhat nebulous control flow concept.

Reply:
After seeing these examples, I agree that it's not the join that's bad, it's the use after join.
It's not just reading the variable which is iffy afterwards, you also shouldn't assign to the variable (because it has no type).

It still means that we need to introduce a *state* after the join which can represent the "unusable" variable.
Maybe just:
* Variable is untyped and definitely unassigned: You can assign to it, and it's an initializing assignment. After the assignment it's not untyped any more, it has an inferred initialization type.
* Variable is untyped and potentially assigned: You cannot read or write the variable. Not even if it's late. You can't lose "potentially assigned" by joining.

Reply:
> With `final`, we don't make it an error to join two paths where one is definitely assigned and the other is not.

One of us is confused.  Again, you need to stop talking about definite assignment, but assuming you mean "join two paths where one has an IIT and one does not", I believe my spec is intended to make this an error:

```
  - If the variable is declared `final` and not `late` it is an error, and there
    is no **initialization inferred type** after the join.
```

Reply:
> After seeing these examples, I agree that it's not the join that's bad, it's the use after join.
> I

I can probably work something out around this.

Reply:
> It still means that we need to introduce a _state_ after the join which can represent the "unusable" variable.

Thinking about this a bit, an unpleasant consequence of this is that we need to take this state into account in joins.  The following program shouldn't be legal:

```
var x;
if (b) {
  if (b) { 
    x = 3.0;
  } else {
   x = 3;
  } // x is joined, and loses its IIT
} else {
  x = "hello";
} // x has not IIT on one path, and an IIT of `String` on another.
  // As currently specified, we would infer `String?` which is wrongo.
  // So we need to track the fact that `x` had an IIT at some point
  // in the past.  It might be sufficient to use "potentially assigned" as 
  // a proxy for this.  Complicated....

Reply:
> I lean slightly towards making the use of the variable the error and not the join itself. Pragmatically, that allows more programs that don't do anything harmful. And I think it draws the user's attention to a concrete entity — a variable they declared that they are using — instead of to a somewhat nebulous control flow concept.

Just a comment, as I run through updating the test cases.  This isn't how it ends up feeling to me.  When the error is at the join point, there's a clear error you can report: I can't infer a type for this variable.  When the error is at the use point, it feels weirder.  What do you say?  "This variable can't be used because I couldn't infer a type" is the best I could get.  You can say "Type can't be inferred for a variable", but that's going to be confusing because I had code that worked until I added a use of a variable.  The user says "you could infer it before, why not now?".

Reply:
I added a patch set implementing this.  You can see the changes from that commit only here: https://github.com/dart-lang/language/pull/1009/commits/ee0d27dc1038e021d72b4cb59df573ba46e0bd44

Reply:
Considering the complexity of working with code that relies on these rules, I agree that reporting the error at the join is preferable.

Otherwise we'll have a situation where any number of variables could have a broken type for any number of lines, and you can't readily see how far back you need to go to find (1) the join where the problem can first be detected, and (2) the predecessors in the flow graph where some unintended IIT caused the problem. The developer needs to find that second location, and we shouldn't make that task intractable.